### DISCUSSION TITLE
Allow user settings to permanently enable "Web search"

### DISCUSSION THREAD

**Discussion by ghost** [open-webui/open-webui#3390] [Ideas]
I would like to set "web search" feature as default for every query to the LLM (this works closer to the way I use perplexity). Every time I have to enable it, I would like the option to set this enabled in my user profile/settings as default.

![image](https://github.com/open-webui/open-webui/assets/144032090/7e7744f5-8c62-4f40-a5cd-11cce187d355)

**Comment by openmoto** [comment]
I'm interested in this as well

**Comment by justinh-rahb** [comment]
https://openwebui.com/f/hub/autotool_filter
+
https://openwebui.com/t/constliakos/web_search

**Reply by dillfrescott** [reply]
Sadly these are kind of buggy. It would be much easier to have an option to enable the built in web search by default.

**Comment by gitfabianmeyer** [comment]
Any updates on this?

**Comment by mdbook** [comment]
Any updates on this?

**Reply by jars101** [reply]
Interested as well

**Comment by hurxxxx** [comment]
### 1. Adding a URL Parameter

By appending a parameter to the URL, we can enable web search functionality:

```
https://localhost:8080/?web-search=true
```
However, this approach doesn’t work when loading an existing conversation.


### 2. Using a Filter Function

Below is an example of a filter function I’ve implemented, which can be applied globally:

```python
from pydantic import BaseModel, Field
from typing import Callable, Awaitable, Any, Optional
from open_webui.utils.middleware import chat_web_search_handler
from open_webui.models.users import UserModel


class Filter:
    class Valves(BaseModel):
        status: bool = Field(default=True)


    async def emit_status(
        self,
        __event_emitter__: Callable[[dict], Awaitable[None]],
        level: str,
        message: str,
        done: bool,
    ):
        if self.valves.status:
            await __event_emitter__(
                {
                    "type": level,
                    "data": {
                        "description": message,
                        "done": done,
                    },
                }
            )

        
    def __init__(self):
        self.valves = self.Valves()

    async def inlet(
        self,
        body: dict,
        __event_emitter__: Callable[[Any], Awaitable[None]],
        __request__: Any,
        __user__: Optional[dict] = None,
        __model__: Optional[dict] = None,
    ) -> dict:
        try:
            user_data = __user__.copy()
            user_data.update({
                "profile_image_url": "",
                "last_active_at": 0,
                "updated_at": 0,
                "created_at": 0
            })

            user_object = UserModel(**user_data)

            if not body.get("features", {}).get("web_search", False):
                await self.emit_status(__event_emitter__, "info", "Web search enabled", True)
                await chat_web_search_handler(__request__, body, {"__event_emitter__": __event_emitter__}, user_object)
                
        except Exception as e:
            print(e)
        return body


``` 

### 3. If I get some time, I’ll try creating a PR to add a user setting that keeps web search always on by default.

**Reply by Mavyre** [reply]
Thank you for this! I'm currently rewritting it to ask the LLM if the user query requires to enable web search before answering.
Do you mind if I base my filter on your code?

Also, do you know why we need to update the user_data like this to pass it to `chat_web_search_handler`? I've tried passing directly __user__ and it breaks the query generation 🤔

**Reply by hurxxxx** [reply]
Feel free to use it as much as you like. The user object received as an argument in the filter does not contain all the necessary attributes required by chat_web_search_handler, so I manually added the required attributes as a workaround.

You may also refer to the following link, where a filter is implemented to automatically choose between web search and knowledge base.

https://github.com/hurxxxx/open-webui-pilelines/blob/main/filters/auto_knowledge_selection.py

**Reply by Mavyre** [reply]
> Feel free to use it as much as you like.

Thank you! It works beautifully, to be honest. I'll publish the Filter to openwebui.com soon :)

> The user object received as an argument in the filter does not contain all the necessary attributes required by chat_web_search_handler, so I manually added the required attributes as a workaround.

Inspired from other filters from openwebui.com, I managed to get the full attributes with this one line: `user = Users.get_user_by_id(__user__["id"])` :)

> You may also refer to the following link, where a filter is implemented to automatically choose between web search and knowledge base.
> 
> https://github.com/hurxxxx/open-webui-pilelines/blob/main/filters/auto_knowledge_selection.py

Thank you again. Looks like it's the same approach I had with auto Web Search and auto Image Generation

**Reply by hurxxxx** [reply]
```
user = Users.get_user_by_id(__user__["id"]) 
```
I didn't use this code due to security concerns. There is no need to include unnecessary information in the user object when sending it. For example, api_key, settings, etc.

**Comment by hurxxxx** [comment]
I added an Always-On Web Search option to User Settings. When turned on, it ensures web search is always enabled. Does this approach make sense?

[always_on_web_search.webm](https://github.com/user-attachments/assets/aecfd5d5-4966-4bb8-85c1-66abace19589)

**Reply by jars101** [reply]
Hello, it will be ideal if this could also be available from Settings > Admin Settings > Interface so admin can enable globally to all users by default.

**Reply by Classic298** [reply]
That is cool. Do a PR!

**Reply by hurxxxx** [reply]
It's a personalization setting, not a global setting under Admin Settings. :)

**Reply by g0t4** [reply]
Why not let it toggle off too? So it is on by default. With `always` it doesn't let you turn it off then.

**Comment by hurxxxx** [comment]
#9370 Check it out

**Reply by tristandostaler** [reply]
It could be nice to have image generation and code interpreter enabled also, no?

**Reply by tristandostaler** [reply]
Well I just tested with both image generation and websearch enabled and it generated an image as well as searched the web lol. Maybe not the best idea for now.

**Reply by hurxxxx** [reply]
previous search might have needed it since the menu was not exposed, but now that the feature is visible, the always-on activation may not be strictly necessary

**Comment by barxilly** [comment]
This is definitely _a step_ in the right direction, but things like custom tools and image searches are still buried away and have to be re-enabled every time

**Comment by dmisdm** [comment]
Not-pretty but working solution if anyone needs it:

Build the Docker image yourself and override the `index.html` with a line like `COPY ./index.html /app/build/index.html`.
With this, I was able to inject an analytics/tracking script, and the following to always re-enable a particular tool upon refresh.

```html
<script>
      let toolId = "your-openwebui-tool-id";
      try {
        let c = JSON.parse(localStorage.getItem("chat-input"));
        if (!c || !c.selectedToolIds) {
          c = {
            prompt: "",
            files: [],
            selectedToolIds: [],
            selectedFilterIds: [],
            imageGenerationEnabled: false,
            webSearchEnabled: false,
            codeInterpreterEnabled: false,
          };
        }
        if (!c.selectedToolIds.includes(toolId)) {
          c.selectedToolIds.push(toolId);
        }
        localStorage.setItem("chat-input", JSON.stringify(c));
      } catch (e) {
        // If parsing/anything fails, just override with default
        try {
          localStorage.setItem(
            "chat-input",
            JSON.stringify({
              prompt: "",
              files: [],
              selectedToolIds: [toolId],
              selectedFilterIds: [],
              imageGenerationEnabled: false,
              webSearchEnabled: false,
              codeInterpreterEnabled: false,
            })
          );
        } catch (e) {
           // Do something here, optionally
        }
      }
    </script>
```

**Comment by Classic298** [comment]
Interface settings provide a toggle to enable web search always

**Reply by filiptrplan** [reply]
This is only for the web search and not all the tools.

**Comment by ForceConstant** [comment]
I too would like this, and saw "Web Search in Chat" in settings and got excited, but it didn't seem to change anything. Web search toggle is still off by default.

Edit: It seems that the problem is just the toggle doesn't show correct state, and web_search is enabled with this setting.

**Reply by Classic298** [reply]
yes correct. you can also enable web search in the model capabilities to always be on
